cloudinit/config/cc_kernel_modules.py

```diff
@@ -0,0 +1,396 @@
+            "alias",
+            "install",
+            "options",
+            "remove",
+        ):
+            if not isinstance(value, str):
+                errors.append(
+                    "Expected a string for kernel_modules:"
+                    f"persist:{key}. Found {value}."
+                )
+        elif key == "blacklist":
+            if not isinstance(value, bool):
+                errors.append(
+                    "Expected a boolean for kernel_modules:"
+                    f"persist:{key}. Found {value}."
+                )
+        elif key == "softdep":
+            for sdkey, sdvalue in sorted(persist[key].items()):
+                if sdkey in ("pre", "post"):
+                    if not isinstance(sdvalue, array):
+                        errors.append(
+                            "Expected an array for kernel_modules:persist:"
+                            f"softdep:{sdkey}. Found {sdvalue}."
+                        )
+
+    return errors
+
+
+def supplemental_schema_validation(kernel_module: dict):
+    """Validate user-provided kernel_modules option values.
+
+    This function supplements flexible jsonschema validation with specific
+    value checks to aid in triage of invalid user-provided configuration.
+
+    @param kernel_module: Dictionary.
+
+    @raises: ValueError describing invalid values provided.
+    """
+    errors = []
+    missing = REQUIRED_KERNEL_MODULES_KEYS.difference(
+        set(kernel_module.keys())
+    )
+    if missing:
+        keys = ", ".join(sorted(missing))
+        errors.append(f"Missing required kernel_modules keys: {keys}")
+
+    for key, value in sorted(kernel_module.items()):
+        if key == "name":
+            if not isinstance(value, str):
+                errors.append(
+                    "Expected a string for kernel_modules:"
+                    f"{key}. Found {value}."
+                )
+        elif key == "load":
+            if not isinstance(value, bool):
+                errors.append(
+                    "Expected a boolean for kernel_modules:"
+                    f"{key}. Found {value}."
+                )
+        elif key == "persist":
+            errors += persist_schema_validation(kernel_module[key])
+
+    if errors:
+        raise ValueError(
+            f"Invalid kernel_modules configuration:{NL}{NL.join(errors)}"
+        )
+
+
+def prepare_module(module_name: str):
+    """Prepare kernel module to be loaded on boot.
+
+    This function appends a kernel module to a file
+    (depends on operating system) for loading a module on boot.
+
+    @param module_name: string.
+
+    @raises: RuntimeError when write operation fails.
+    """
+
+    LOG.info(
+        "Writing %s file for loading kernel modules on boot",
+        DEFAULT_CONFIG["km_files"]["load"].get("path"),
+    )
+
+    try:
+        LOG.debug("Appending kernel module %s", module_name)
+        util.write_file(
+            DEFAULT_CONFIG["km_files"]["load"]["path"],
```

Comment:
In theory we could pull this attribute off of the distro class so that other distros could override that behavior appropriately if available.